queues.lisp

```diff
@@ -0,0 +1,67 @@
+;;;; Erlangen queue arsenal, these queues are multi-producer, single-consumer
+;;;; FIFO queues.
+
+(in-package :erlangen.queues)
+
+;;; Intrusive Vyukov queue:
+
+(defstruct (queue (:constructor make-queue%))
+  (head (error "Need HEAD") :type cons)
+  (tail (error "Need TAIL") :type cons))
+
+(defun make-queue (&aux (stub (list nil)))
+  (make-queue% :head stub :tail stub))
+
+(defun queue-push (queue node)
+  (setf (cdr (xchg (queue-head queue) node)) node))
+
+(defun queue-pop (queue)
+  (let* ((tail (queue-tail queue))
+         (next (cdr tail)))
+    (when next
+      (setf (queue-tail queue) next))))
+
+;;; Intrusive bounded queue, fails on overflow:
+
+(defstruct cell
+  (sequence (error "Need SEQUENCE") :type fixnum)
+  value)
+
+(defstruct (bounded-queue (:constructor make-bounded-queue%))
+  (buffer (error "Need BUFFER") :type (simple-array cell (*)))
+  (mask (error "Need MASK") :type fixnum)
+  (read 0 :type fixnum)
+  (write 0 :type fixnum))
+
+(defun make-bounded-queue (size)
+  (check-type size (and (integer 2) fixnum))
+  (assert (= (logand size (1- size)) 0))
```

Comment:
- [ ] should document that `:mailbox-size` needs to be a power of two. (Also add a telling assertion error here!!)